shared/src/types/ethereum_events/vote_extensions.rs

```diff
@@ -0,0 +1,229 @@
+    where
+        D: DB + for<'iter> DBIter<'iter> + Sync + 'static,
+        H: StorageHasher + Sync + 'static,
+    {
+        let VoteExtensionDigest {
+            signatures,
+            events,
+            nulls,
+        } = digest;
+
+        let mut extensions = vec![];
+
+        for (sig, addr) in signatures.into_iter() {
+            let mut ext = VoteExtension {
+                block_height: self.storage.last_height,
+                events: vec![],
+            };
+
+            let ext = if nulls.contains(&addr) {
+                ext
+            } else {
+                for event in events {
+                    if event.signers.contains(&addr) {
+                        ext.events.push(event.clone());
+                    }
+                }
+                ext.events.sort();
+            };
+
+            let signed = Signed {
+                data: ext,
+                sig,
+            };
+            extensions.push(sig);
+        }
+        extensions
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::types::ethereum_events::Uint;
+    use crate::types::hash::Hash;
+
+    /// Test the hashing of an Ethereum event
+    #[test]
+    fn test_ethereum_event_hash() {
+        let nonce = Uint::from(123u64);
+        let event = EthereumEvent::TransfersToNamada {
+            nonce,
+            transfers: vec![],
+        };
+        let hash = event.hash().unwrap();
+
+        assert_eq!(
+            hash,
+            Hash([
+                94, 131, 116, 129, 41, 204, 178, 144, 24, 8, 185, 16, 103,
+                236, 209, 191, 20, 89, 145, 17, 41, 233, 31, 98, 185, 6,
+                217, 204, 80, 38, 224, 23
+            ])
+        );
+    }
+
+    /// This test is ultimately just exercising the underlying
+    /// library we use for fractions, we want to make sure
+    /// operators work as expected with our FractionalVotingPower
+    /// type itself
+    #[test]
+    fn test_fractional_voting_power_ord_eq() {
+        assert!(
+            FractionalVotingPower::new(2, 3).unwrap()
+                > FractionalVotingPower::new(1, 4).unwrap()
+        );
+        assert!(
+            FractionalVotingPower::new(1, 3).unwrap()
+                > FractionalVotingPower::new(1, 4).unwrap()
+        );
+        assert!(
+            FractionalVotingPower::new(1, 3).unwrap()
+                == FractionalVotingPower::new(2, 6).unwrap()
+        );
+    }
+
+    /// Test error handling on the FractionalVotingPower type
+    #[test]
+    fn test_fractional_voting_power_valid_fractions() {
+        assert!(FractionalVotingPower::new(0, 0).is_err());
+        assert!(FractionalVotingPower::new(1, 0).is_err());
+        assert!(FractionalVotingPower::new(0, 1).is_ok());
+        assert!(FractionalVotingPower::new(1, 1).is_ok());
+        assert!(FractionalVotingPower::new(1, 2).is_ok());
+        assert!(FractionalVotingPower::new(3, 2).is_err());
+    }
```

Comment:
I would love a unit test or two for the decompress algorithm if that's not too much trouble.